### PR TITLE
Add database implementation of  SemesterRepository

### PR DESCRIPTION
**Pull request by tenorz007**

### CODE REVIEW

**Comment by tenorz007** [comment]
Completes Semester in #30

**Comment by tenorz007** [comment]
Approved!

**Comment by BrainMaestro** [comment]
Remember. This should be squashed to one commit

**Comment by tenorz007** [comment]
Naaa. I prefer to see the changes. It's better that way.

**Comment by BrainMaestro** [comment]
It's not. This is one logical change. Should not be split over several commits. Just pollutes the history

File: app/Repositories/Database/SQLSemesterRepository.php [hunk 1]
```diff
@@ -0,0 +1,79 @@
...
+    /**
+     * Set current semester by its id
+     *
+     * @param $id
+     * @return boolean
+     */
+    public function setCurrentSemester($id)
+    {
+        $currentSemester = $this->find($id);
+        $currentSemester['data']['current_semester'] = true;
```
Comment (BrainMaestro) [inline app/Repositories/Database/SQLSemesterRepository.php]: This is redundant. The update method already finds the model. Check any of the methods I wrote that use the update

File: app/Transformers/SemesterTransformer.php [hunk 2]
```diff
@@ -20,9 +20,10 @@ class SemesterTransformer extends TransformerAbstract
     public function transform(Semester $semester)
     {
         return [
-            'id' => (int) $semester->id,
-            'semester' => $semester->semester,
-            'current_semester' => (boolean) $semester->current_semester,
+            'id' => (int)$semester->id,
```
Comment (BrainMaestro) [inline app/Transformers/SemesterTransformer.php]: This is not psr2 compliant. There should be a space between the cast operator and the variables

File: tests/unit/Repositories/Database/SQLSemesterRepositoryTest.php [hunk 3]
```diff
@@ -0,0 +1,83 @@
...
+
+        $semesters = self::$semesterRepository->getSemesters();
+
+        $this->assertCount(count($createdSemesters), $semesters['data']);
+        $this->assertEquals($createdSemesters, $semesters['data']);
+    }
+
+    public function testSetCurrentSemester()
+    {
+        factory(Fce\Models\Semester::class, 2)->create();
```
Comment (BrainMaestro) [inline tests/unit/Repositories/Database/SQLSemesterRepositoryTest.php]: Not necessary

File: tests/unit/Repositories/Database/SQLSemesterRepositoryTest.php [hunk 4]
```diff
@@ -0,0 +1,83 @@
...
+        $semester = self::$semesterRepository->setCurrentSemester($this->semester->id);
+
+        $this->assertTrue($semester);
```
Comment (BrainMaestro) [inline tests/unit/Repositories/Database/SQLSemesterRepositoryTest.php]: You should be asserting that the field you updated was actually changed. This doesn't do that

File: app/Transformers/SemesterTransformer.php (hunk 2 above)
Comment (tenorz007) [inline app/Transformers/SemesterTransformer.php]: Trying using the automatic formatting and see what happens.

File: tests/unit/Repositories/Database/SQLSemesterRepositoryTest.php [hunk 5]
```diff
@@ -0,0 +1,83 @@
...
+    }
+
+    /**
+     * @depends testSetCurrentSemester
+     */
+    public function testGetCurrentSemester()
```
Comment (BrainMaestro) [inline tests/unit/Repositories/Database/SQLSemesterRepositoryTest.php]: This should probably also assert that only one result was returned

File: tests/unit/Repositories/Database/SQLSemesterRepositoryTest.php [hunk 6]
```diff
@@ -0,0 +1,83 @@
...
+        factory(Fce\Models\Semester::class, 2)->create();
+        self::$semesterRepository->setCurrentSemester($this->semester->id);
+
+        $this->semester  = self::$semesterRepository->getCurrentSemester();
+
+        $this->assertTrue($this->semester['data']['current_semester']);
+    }
+
+    /**
+     * @depends testGetCurrentSemester
```
Comment (BrainMaestro) [inline tests/unit/Repositories/Database/SQLSemesterRepositoryTest.php]: This is for when you want to pass a return value between tests. You're not doing that, so you should probably remove it

File: tests/unit/Repositories/Database/SQLSemesterRepositoryTest.php (hunk 3 above)
Comment (tenorz007) [inline tests/unit/Repositories/Database/SQLSemesterRepositoryTest.php]: It is

File: tests/unit/Repositories/Database/SQLSemesterRepositoryTest.php (hunk 6 above)
Comment (BrainMaestro) [inline tests/unit/Repositories/Database/SQLSemesterRepositoryTest.php]: Also I can see two spaces after `$this->semester`

File: tests/unit/Repositories/Database/SQLSemesterRepositoryTest.php (hunk 4 above)
Comment (tenorz007) [inline tests/unit/Repositories/Database/SQLSemesterRepositoryTest.php]: You'll have to talk to your update method then cause it only returns true.

File: tests/unit/Repositories/Database/SQLSemesterRepositoryTest.php (hunk 5 above)
Comment (tenorz007) [inline tests/unit/Repositories/Database/SQLSemesterRepositoryTest.php]: No need as it returns an item since, one is passed. If you were to use count you would use ['data'] where this would return the number or rows which would be wrong, and it you choose to do without data, it won't be consistent with other tests.

File: tests/unit/Repositories/Database/SQLSemesterRepositoryTest.php (hunk 6 above)
Comment (tenorz007) [inline tests/unit/Repositories/Database/SQLSemesterRepositoryTest.php]: Nope, if the set function doesn't pass, there is no need to try the get because the it depends on the set actually working, so if the set test fails then there is no need to carry on.

File: app/Transformers/SemesterTransformer.php (hunk 2 above)
Comment (BrainMaestro) [inline app/Transformers/SemesterTransformer.php]: It has a space with automatic formatting. Besides, can't you see that it's clearly better the other way.

File: tests/unit/Repositories/Database/SQLSemesterRepositoryTest.php (hunk 3 above)
Comment (BrainMaestro) [inline tests/unit/Repositories/Database/SQLSemesterRepositoryTest.php]: You're creating 2 semesters and you're not using them for anything

File: tests/unit/Repositories/Database/SQLSemesterRepositoryTest.php (hunk 4 above)
Comment (BrainMaestro) [inline tests/unit/Repositories/Database/SQLSemesterRepositoryTest.php]: That's how the update works. Have you even read any of the stuff I did? If you have, you'll know how I tested updates.

File: app/Models/QuestionSet.php [hunk 7]
```diff
@@ -22,15 +22,4 @@ public function questions()
...
         return $this->belongsToMany(Question::class);
     }
-
-    /**
-     * The QuestionSet relationship to Semester
-     * A questionSet belongsToMany semesters
-     *
-     * @return \Illuminate\Database\Eloquent\Relations\BelongsToMany
-     */
-    public function semesters()
```
Comment (BrainMaestro) [inline app/Models/QuestionSet.php]: What? Why did you remove the relationships

Comment (tenorz007) [inline app/Models/QuestionSet.php]: Cause it is not needed

Comment (BrainMaestro) [inline app/Models/QuestionSet.php]: Think about it carefully. How would we know which question sets are being used for the semester.


Sent from Outlook Mobile






On Wed, Feb 24, 2016 at 1:22 AM -0800, "tenorz007" <notifications@github.com> wrote:





> @@ -22,15 +22,4 @@ public function questions()
>      {
>          return $this->belongsToMany(Question::class);
>      }
> -
> -    /**
> -     * The QuestionSet relationship to Semester
> -     * A questionSet belongsToMany semesters
> -     *
> -     * @return \Illuminate\Database\Eloquent\Relations\BelongsToMany
> -     */
> -    public function semesters()

Cause it is not needed

---
Reply to this email directly or view it on GitHub:
https://github.com/BrainMaestro/FCE/pull/41/files#r53912017

Comment (tenorz007) [inline app/Models/QuestionSet.php]: Check the migration, there was no foreign key in either. That's why I
removed it.

Comment (BrainMaestro) [inline app/Models/QuestionSet.php]: It's a many to many relationship. There's a pivot table migration.


Sent from Outlook Mobile






On Wed, Feb 24, 2016 at 3:12 AM -0800, "tenorz007" <notifications@github.com> wrote:





> @@ -22,15 +22,4 @@ public function questions()
>      {
>          return $this->belongsToMany(Question::class);
>      }
> -
> -    /**
> -     * The QuestionSet relationship to Semester
> -     * A questionSet belongsToMany semesters
> -     *
> -     * @return \Illuminate\Database\Eloquent\Relations\BelongsToMany
> -     */
> -    public function semesters()

Check the migration, there was no foreign key in either. That's why I
removed it.
On Feb 24, 2016 12:09, "BrainMaestro" <notifications@github.com> wrote:

> In app/Models/QuestionSet.php
> <https://github.com/BrainMaestro/FCE/pull/41#discussion_r53923781>:
>
> > @@ -22,15 +22,4 @@ public function questions()
> >      {
> >          return $this->belongsToMany(Question::class);
> >      }
> > -
> > -    /**
> > -     * The QuestionSet relationship to Semester
> > -     * A questionSet belongsToMany semesters
> > -     *
> > -     * @return \Illuminate\Database\Eloquent\Relations\BelongsToMany
> > -     */
> > -    public function semesters()
>
> Think about it carefully. How would we know which question sets are being
> used for the semester. Sent from Outlook Mobile
> On Wed, Feb 24, 2016 at 1:22 AM -0800, "tenorz007" <
> notifications@github.com> wrote: @@ -22,15 +22,4 @@ public function
> questions() { return $this->belongsToMany(Question::class); } - - /** - *
> The QuestionSet relationship to Semester - * A questionSet belongsToMany
> semesters - * - * @return <https://github.com/return>
> \Illuminate\Database\Eloquent\Relations\BelongsToMany - */ - public
> function semesters()
> Cause it is not needed
> … <#-706495963_>
> --- Reply to this email directly or view it on GitHub:
> https://github.com/BrainMaestro/FCE/pull/41/files#r53912017
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/BrainMaestro/FCE/pull/41/files#r53923781>.
>


---
Reply to this email directly or view it on GitHub:
https://github.com/BrainMaestro/FCE/pull/41/files#r53924078

File: app/Transformers/SemesterTransformer.php (hunk 2 above)
Comment (tenorz007) [inline app/Transformers/SemesterTransformer.php]: https://github.com/php-fig-rectified/fig-rectified-standards/blob/master/PSR-2-R-coding-style-guide-additions.md @BrainMaestro

File: tests/unit/Repositories/Database/SQLSemesterRepositoryTest.php (hunk 3 above)
Comment (tenorz007) [inline tests/unit/Repositories/Database/SQLSemesterRepositoryTest.php]: I am using it, to ensure that there are other things in the database, and only that one was changed.

File: app/Models/QuestionSet.php (hunk 7 above)
Comment (tenorz007) [inline app/Models/QuestionSet.php]: Will revert

File: app/Transformers/SemesterTransformer.php (hunk 2 above)
Comment (BrainMaestro) [inline app/Transformers/SemesterTransformer.php]: I saw that. It doesn't say anything about the spacing. Just that typecasts should be used. Check everywhere else. Even how Phil Sturgeon himself writes transformers.


I'll change it if you don't.


Sent from Outlook Mobile






On Wed, Feb 24, 2016 at 4:03 AM -0800, "tenorz007" <notifications@github.com> wrote:





> @@ -20,9 +20,10 @@ class SemesterTransformer extends TransformerAbstract
>      public function transform(Semester $semester)
>      {
>          return [
> -            'id' => (int) $semester->id,
> -            'semester' => $semester->semester,
> -            'current_semester' => (boolean) $semester->current_semester,
> +            'id' => (int)$semester->id,

https://github.com/php-fig-rectified/fig-rectified-standards/blob/master/PSR-2-R-coding-style-guide-additions.md @BrainMaestro

---
Reply to this email directly or view it on GitHub:
https://github.com/BrainMaestro/FCE/pull/41/files#r53928926

File: tests/unit/Repositories/Database/SQLSemesterRepositoryTest.php (hunk 3 above)
Comment (BrainMaestro) [inline tests/unit/Repositories/Database/SQLSemesterRepositoryTest.php]: OK. Then maybe you need an extra check to be sure that those ones weren't affected


Sent from Outlook Mobile






On Wed, Feb 24, 2016 at 4:03 AM -0800, "tenorz007" <notifications@github.com> wrote:





> +    {
> +        $createdSemesters = factory(Fce\Models\Semester::class, 2)->create();
> +        $createdSemesters = array_merge(
> +            [SQLSemesterRepository::transform($this->semester)['data']],
> +            SQLSemesterRepository::transform($createdSemesters)['data']
> +        );
> +
> +        $semesters = self::$semesterRepository->getSemesters();
> +
> +        $this->assertCount(count($createdSemesters), $semesters['data']);
> +        $this->assertEquals($createdSemesters, $semesters['data']);
> +    }
> +
> +    public function testSetCurrentSemester()
> +    {
> +        factory(Fce\Models\Semester::class, 2)->create();

I am using it, to ensure that there are other things in the database, and only that one was changed.

---
Reply to this email directly or view it on GitHub:
https://github.com/BrainMaestro/FCE/pull/41/files#r53928972

**Comment by BrainMaestro** [comment]
Approved
